Update Span merge

naoman opened the pull request:
In the current implementation, when we merge spans from client and server, we compute timestamp and duration for the Span by looking at the min and max timestamps. This breaks if we have a clock skew. Once timestamp and duration is set here, we don't recompute it in [ApplyTimestampAndDuration](https://github.com/openzipkin/zipkin/blob/1d0e657e6f89b929a0be27db843a848611242b92/zipkin/src/main/java/zipkin/internal/ApplyTimestampAndDuration.java#L34). 

In this change all we are doing is set timestamp and duration to null, when we merge two spans and and both have direct timestamp set.

naoman committed:
Updating how we compute span duration and timestamp when we merge partial spans together, to avoid issues when there is a clock skew between client and server

naoman committed:
updating comment

naoman committed:
updating comment

codefromthecrypt commented on zipkin/src/test/java/zipkin/storage/SpanStoreTest.java:
```diff
@@ -806,6 +806,43 @@ public void clientTimestampAndDurationWinInSharedSpan() {
     }
   }
 
+  /**
```
I'd move this test to SpanTest as it doesn't require live servers to verify

codefromthecrypt commented on zipkin/src/test/java/zipkin/storage/SpanStoreTest.java:
```diff
@@ -789,7 +789,7 @@ public void clientTimestampAndDurationWinInSharedSpan() {
     // neither client, nor server set span.timestamp, duration
     Span clientViewDerived = Span.builder().traceId(1).name("derived").id(666)
         .addAnnotation(Annotation.create(clientTimestamp, CLIENT_SEND, client))
-        .addAnnotation(Annotation.create(clientTimestamp + clientDuration, CLIENT_SEND, client))
```
keep this fix though!

codefromthecrypt commented on zipkin/src/main/java/zipkin/Span.java:
```diff
@@ -253,11 +253,11 @@ public Builder merge(Span that) {
-        this.timestamp = Math.min(this.timestamp, that.timestamp);
-        this.duration = Math.max(thisEndTs, thatEndTs) - this.timestamp;
+      } else {
+        // We have 2 different timestamps and we don't know which one is authoritative.
```
we actually do. the client is authoritative in a shared span. Let's set it to the client one (span with "cr" annotation), or null if cr is absent

codefromthecrypt replied:
so a suggested fix (which considers client span) would be to move this block below where we iterate through annotations. When we iterate through annotations decide which side is the client span and prefer that somehow..

codefromthecrypt commented:
thanks for the help!

naoman committed:
Refactoring code, recommendations from pull request review

naoman commented:
Thanks for reviewing the code @adriancole. I've made following changes:
1. Moved the test code from SpanStoreTest to SpanTest
2. Updated Span.Builder.merge() code to return timestamp and duration of client after merge. However, instead of computing it by going through annotations, I'm calling ApplyTimestampAndDuration.apply(). Since the code is so similar, I didn't want to copy paste. And later if we want to change the logic of computing ApplyTimestampAndDuration, there will be just one place to make the change. Let me know what you think.

codefromthecrypt commented on zipkin/src/main/java/zipkin/Span.java:
```diff
@@ -253,22 +264,13 @@ public Builder merge(Span that) {
+        // We have 2 different timestamps. Use ApplyTimestampAndDuration to compute final timestamp.
+        this.timestamp = null;
+        this.duration = null;
+        return ApplyTimestampAndDuration.apply(this.build()).toBuilder();
```
we can't do this as we only include authoritative data reported to us (as opposed to deriving span's timestamp from annotations). The prior data incorrectly tried to merge authoritatively set timestamps (which was the bug). The better way is to prefer the client as mentioned before.

naoman committed:
Keep authoritative data only, and not re-computing it

codefromthecrypt commented:
Thanks for the help on this. I appreciate you've spent some time on this, so if you don't have time to budget to loop through the annotations once, I can do it as a part of merging. lemme know!

naoman commented:
It would be great if you can do it as part of the merge. I couldn't think of a cleaner way of doing this, since we were only looping through "that" annotations before the change.
Thanks!

codefromthecrypt commented:
@naoman thanks for the help. I added a little temporary debt because what you are trying to sort out will be straightforward in the next model. https://github.com/openzipkin/zipkin/commit/80e99c482e06da7fa0fdc8cbf1055d07bb690853
